### PR TITLE
Mechanism for rememembering dependency versions

### PR DESCRIPTION
**Pull request by dmdashenkov**
Often, when developing Gradle plugins, we need the versions of this or that Maven artifact at plugin runtime.
To retrieve those versions, we store them into a file at build time and place that file into the plugin runtime classpath.
Then, at plugin runtime, we use `DependncyVersions` to extract the versions from the resource file.

In this PR we expand the existing `DependncyVersions` mechanism to include not only the predefined artifacts, but any version for any artifact. There can also now be many versions files.

### CODE REVIEW

File: plugin-base/src/main/java/io/spine/tools/gradle/DependencyVersions.java [hunk 1]
```diff
@@ -26,103 +26,74 @@
...
-
-        Property(String name) {
-            this.name = name;
-        }
-
-        private String from(Properties properties) {
-            String value = properties.getProperty(name);
-            checkNotNull(value, "The property `%s` could not be found.", name);
-            return value;
+    public Optional<String> versionOf(Dependency dependency) {
```
Comment (yuri-sergiichuk) [inline plugin-base/src/main/java/io/spine/tools/gradle/DependencyVersions.java:91]: Let's add a `checkNotNull` while the method is `public`.

**Review by yuri-sergiichuk — COMMENTED** [review]
@dmdashenkov overall LGTM. Can we have some tests for the new behavior? The code is full of CodeCov check warnings about not being covered with tests.

**Comment by dmdashenkov** [comment]
@yuri-sergiichuk, PTAL again.

File: plugin-base/src/main/java/io/spine/tools/gradle/Dependency.java [hunk 2]
```diff
@@ -45,6 +46,25 @@
...
+     */
+    default String fileSafeId() {
+        return groupId() + '_' + name();
+    }
+
+    /**
+     * Compiles an {@link Artifact} for this dependency with a version from the given
+     * set of versions.
+     */
+    default Artifact withVersionFrom(DependencyVersions versions) {
```
Comment (yuri-sergiichuk) [inline plugin-base/src/main/java/io/spine/tools/gradle/Dependency.java:60]: Please add a `checkNotNull` check here.

File: plugin-base/src/main/java/io/spine/tools/gradle/ProtocConfigurationPlugin.java [hunk 3]
```diff
@@ -50,9 +51,13 @@
...
  * no action is performed.
  */
+@SuppressWarnings("AbstractClassNeverImplemented")
+    // Implemented in language-specific parts of Model Compiler.
 public abstract class ProtocConfigurationPlugin extends SpinePlugin {
 
-    protected static final DependencyVersions VERSIONS = DependencyVersions.get();
+    @SuppressWarnings("DuplicateStringLiteralInspection") // A duplicate in tests.
+    @VisibleForTesting
+    static final DependencyVersions versions = DependencyVersions.loadFor("plugin-base");
```
Comment (yuri-sergiichuk) [inline plugin-base/src/main/java/io/spine/tools/gradle/ProtocConfigurationPlugin.java:60]:
```suggestion
    @VisibleForTesting
    @SuppressWarnings("DuplicateStringLiteralInspection" /* A duplicate in tests. */)
    static final DependencyVersions versions = DependencyVersions.loadFor("plugin-base");
```

**Review by yuri-sergiichuk — APPROVED** [review]

**Pull request merged**